### PR TITLE
Retry PoolNeedsConnection Errors

### PR DESCRIPTION
**Pull request by nplasterer**
Makes `PoolNeedsConnection` Errors retryable for actions that require backgrounding the app while doing an action like receiving a signature on adding and revoking installations and applying a signature request.

Also adds a new error type that can be caught for the group create issue.

Fixes https://github.com/xmtp/libxmtp/issues/996

### CODE REVIEW

File: xmtp_mls/src/storage/errors.rs [hunk 1]
```diff
@@ -68,6 +68,7 @@ impl RetryableError for StorageError {
             Self::Pool(_) => true,
             Self::Lock(_) => true,
             Self::SqlCipherNotLoaded => true,
+            Self::PoolNeedsConnection => true,
```
Comment (nplasterer) [inline xmtp_mls/src/storage/errors.rs:71]: This means the error will retry 5 times before erroring. That might be okay? I also took a pass at just doing it in the code and retrying 3 times here https://github.com/xmtp/libxmtp/pull/1010/commits/45adbfe9ee1641694db84a3778a11700e16c5e2d open to others thoughts on how we should handle this.

Comment (neekolas) [reply xmtp_mls/src/storage/errors.rs:71]: This definitely feels like it should be retryable. Thanks for fixing

**Review by alexrisch — APPROVED** [review]

**Comment by insipx** [comment]
Just reposting here for visibility :


Was just thinking about this a little

During these sessions, is the XMTP client still in memory/we're not re-instantiating the struct?

Instead of retrying, which may still fail depending on the threading situation -- sortof feels like we're just waiting for a race which reconnects the db in another thread -- we could instead add a flag on the struct and a boolean passed to release_connection like `reconnect_on_reentry`, that would auto-re-establish a database connection instead of throwing an error. This seems to check if you're already calling to release the db connection. I don't fully understand what the app is back active callback is, though. would be happy to look through the sdk code for that if someone can link it.


also, this retry would only work if the reconnect is already in a codepath that is in a `retry!()` block. We could surround this in another `retry` block, but since this is a function used in lots of other places, may inadvertently cause double-retry times

**Comment by nmalzieu** [comment]
> Just reposting here for visibility :
> 
> Was just thinking about this a little
> 
> During these sessions, is the XMTP client still in memory/we're not re-instantiating the struct?
> 
> Instead of retrying, which may still fail depending on the threading situation -- sortof feels like we're just waiting for a race which reconnects the db in another thread -- we could instead add a flag on the struct and a boolean passed to `release_connection` like `reconnect_on_reentry`, that would auto-re-establish a database connection instead of throwing an error. This seems to check if you're already calling to release the db connection. I don't fully understand what the `app is back active callback` is, though. would be happy to look through the sdk code for that if someone can link it.
> 
> also, this retry would only work if the reconnect is already in a codepath that is in a `retry!()` block. We could surround this in another `retry` block, but since this is a function used in lots of other places, may inadvertently cause double-retry times

Hey @insipx , I'm not 100% sure I get what you mean, but one thing we DON'T want is for libxmtp tu auto-re-establish a database connection. Because libxmtp has no idea if the app is active or backgrounded, and on iOS opening a connection to the libxmtp db when the app gets suspended will lead to a crash

File: xmtp_mls/src/groups/mod.rs [hunk 2]
```diff
@@ -187,6 +187,8 @@ pub enum GroupError {
     SqlKeyStore(#[from] sql_key_store::SqlKeyStoreError),
     #[error("No pending commit found")]
     MissingPendingCommit,
+    #[error("Sync failed to wait for intent: {0}")]
+    SyncFailedToWait(String),
```
Comment (nplasterer) [inline xmtp_mls/src/groups/mod.rs:191]: @neekolas adding this so it's easier to catch the error that is causing group creates to fail.

File: xmtp_mls/src/identity_updates.rs [hunk 3]
```diff
@@ -259,9 +267,14 @@ where
...
-            .get_association_state(&self.store().conn()?, &inbox_id, None)
-            .await?;
+
+        let current_state = retry_async!(
+            Retry::default(),
+            (async {
+                self.get_association_state(&self.store().conn()?, &inbox_id, None)
+                    .await
+            })
+        )?;
```
Comment (nplasterer) [inline xmtp_mls/src/identity_updates.rs:277]: This should cover the case converse is hitting when the database is trying to reconnect after getting a signature from a different app for revoking installations.

File: xmtp_mls/src/groups/mod.rs [hunk 4]
```diff
@@ -188,6 +188,8 @@ pub enum GroupError {
     SqlKeyStore(#[from] sql_key_store::SqlKeyStoreError),
     #[error("No pending commit found")]
     MissingPendingCommit,
+    #[error("Sync failed to wait for intent: {0}")]
```
Comment (insipx) [inline xmtp_mls/src/groups/mod.rs:191]:
```suggestion
    #[error("Sync failed to wait for intent")]
```

File: xmtp_mls/src/groups/mod.rs [hunk 5]
```diff
@@ -188,6 +188,8 @@ pub enum GroupError {
...
+    SyncFailedToWait(String),
```
Comment (insipx) [inline xmtp_mls/src/groups/mod.rs:192]:
```suggestion
    SyncFailedToWait,
```

File: xmtp_mls/src/groups/sync.rs [hunk 6]
```diff
@@ -221,7 +221,9 @@ impl MlsGroup {
             num_attempts += 1;
         }
 
-        Err(last_err.unwrap_or(GroupError::Generic("failed to wait for intent".to_string())))
+        Err(last_err.unwrap_or(GroupError::SyncFailedToWait(
```
Comment (insipx) [inline xmtp_mls/src/groups/sync.rs:224]:
```suggestion
        Err(last_err.unwrap_or(GroupError::SyncFailedToWait)
```

File: xmtp_mls/src/groups/sync.rs [hunk 7]
```diff
@@ -221,7 +221,9 @@ impl MlsGroup {
...
+            "failed to wait for intent".to_string(),
```
Comment (insipx) [inline xmtp_mls/src/groups/sync.rs:225]:
```suggestion
```

File: xmtp_mls/src/groups/sync.rs [hunk 8]
```diff
@@ -221,7 +221,9 @@ impl MlsGroup {
...
+        )))
```
Comment (insipx) [inline xmtp_mls/src/groups/sync.rs:226]:
```suggestion
        ))
```

**Review by insipx — APPROVED** [review]
Looks good!

When writing errors, the `#[error("display impl")]` is the `Display` impl for `to_string`, so an extra string is redundant

**Review by neekolas — APPROVED** [review]

**Pull request merged**